enrich README (#24)
b5g-ex/zenohex: README.md

[![Hex version](https://img.shields.io/hexpm/v/zenohex.svg "Hex version")](https://hex.pm/packages/zenohex)
[![API docs](https://img.shields.io/hexpm/v/zenohex.svg?label=hexdocs "API docs")](https://hexdocs.pm/zenohex/)
[![License](https://img.shields.io/hexpm/l/zenohex.svg)](https://github.com/zenohex/zenohex/blob/main/LICENSE)
[![CI](https://github.com/b5g-ex/zenohex/actions/workflows/ci.yml/badge.svg)](https://github.com/b5g-ex/zenohex/actions/workflows/ci.yml)

> Zenohex is Elixir API for [Zenoh](https://zenoh.io/).

Author: zenoh-pythonなど他のライブラリもこのような書き方（e.g. Python API for Zenoh）なので合わせました．

Zenoh is a new protocol for Zero Overhead Pub/Sub, Store/Query and Compute.
The most obvious explanation is that Zenoh offers publication subscription-based communication capabilities.
Within the same network, Zenoh can autonomously search for communication partner nodes like DDS.
Between different networks, Zenoh can search for nodes through a broker (called a router in Zenoh) like MQTT.
Also, Zenoh provides functions for database operations and computational processing based on the Key-Value Store.
Moreover, it has plugins/bridges for interoperability with MQTT, DDS, REST, etc. as communication middleware, and influxdb, RocksDB, etc. as database stacks.

For more details about Zenoh, please refer to the official resources.
- [Official Page](https://zenoh.io/)
- [GitHub](https://github.com/eclipse-zenoh/zenoh)

> - [Discord](https://discord.gg/vSDSpqnbkm)

Author: ちょっと前書きとしては長いんですが,,, AlchemistがZenoh自体に興味を惹いてもらうために:D

> For more details about Zenoh, please refer to the official resources.
> - [Official Page](https://zenoh.io/)
> - [GitHub](https://github.com/eclipse-zenoh/zenoh)
> - [Discord](https://discord.gg/vSDSpqnbkm)

Reviewer: ```suggestion
For more details about Zenoh, please refer to the official resources.

- [Official Page](https://zenoh.io/)
- [GitHub](https://github.com/eclipse-zenoh/zenoh)
- [Discord](https://discord.gg/vSDSpqnbkm)
```
[prettier](https://github.com/prettier/prettier) の指摘（一行あけるべし）を受け修正

Zenoh's core modules are implemented in Rust, but API libraries in various programming languages such as Python ([zenoh-python](https://github.com/eclipse-zenoh/zenoh-python)), C ([zenoh-c](https://github.com/eclipse-zenoh/zenoh-c)), C++ ([zenoh-cpp](https://github.com/eclipse-zenoh/zenoh-cpp)) are officially provided.

So what we need is [Elixir](https://elixir-lang.org/)!
With this library, you can call Zenoh from your Elixir application to perform its basic processing.

> This allows the creation and communication of a large number of fault-tolerant nodes with little memory load (we hope :D

Author: ちょっと前置きとしては長いんですが,,, 初めてこれを見る alchemist に Zenoh への興味を惹いてもらうために,,,:D

## Usage

**Currently, Zenohex uses version 0.10.1-rc of Zenoh.  
We recommend you use the same version to communicate with other Zenoh clients or routers.**

### Installation

`zenohex` is [available in Hex](https://hex.pm/packages/zenohex).

You can install this package into your project by adding `zenohex` to your list of dependencies in `mix.exs`:

```elixir
  defp deps do
    [
      ...
      {:zenohex, "~> 0.2.0-rc.2"},
      ...
    ]
  end
```

Documentation is also [available in HexDocs](https://hexdocs.pm/zenohex).

Zenohex can be also adapted to your [Nerves](https://nerves-project.org/) application just by adding `zenohex` in `mix.exs`.

> Please refer to [pojiro/nerves_zenohex](https://github.com/pojiro/nerves_zenohex) as the example.

Reviewer: これをやるなら、リポジトリの名前を `zenohex_on_nerves` にして、体裁を整えたいですね。
今のままだとナンノコッチャになりそうなので。

> Please refer to [pojiro/nerves_zenohex](https://github.com/pojiro/nerves_zenohex) as the example.

Author: @pojiro お手数おかけしますが，これでぜひにです！
`mix hex.publish docs --revert VERSION` でドキュメントだけの更新ならいつでもできますし 🙏 
https://hexdocs.pm/hex/Mix.Tasks.Hex.Publish.html#module-reverting-a-package

> Please refer to [pojiro/nerves_zenohex](https://github.com/pojiro/nerves_zenohex) as the example.

Reviewer: ```suggestion
Please refer to [pojiro/zenohex_on_nerves](https://github.com/pojiro/zenohex_on_nerves) as the example.
```
後で取り回すのめんどくさいので、空のリポジトリ作りました。あとで、そこにプッシュして完成させるようにすれば🆗？

> This repository uses [Rustler](https://github.com/rusterlium/rustler) to call Rust (Zenoh) modules from Elixir, and pre-compiled NIF modules are automatically downloaded at `mix compile` time (since v0.1.3). 

Reviewer: ```suggestion
This repository uses [Rustler](https://github.com/rusterlium/rustler) to call Rust (Zenoh) modules from Elixir, and pre-compiled NIF modules are automatically downloaded at `mix compile` time (since v0.1.3).
```

prettier で見つけた不要なスペースを削除

IOW, if you just want to use this library from your Elixir application, you do not need to prepare a Rust environment.

> If you still want to build Rust NIF modules locally, please refer to [this section](#build-nif-module-locally).

Author: 元のだと [optional] を見逃すと Rust が必要そうに読めなくもなかったので，ちょっと明確化しました．

### Getting Started

Zenohex has a policy of providing APIs that wrap the basic functionality of Zenoh like other API libraries.

> Here is the first step to building an Elixir application and using this feature.

Author: このライブラリの方針を明確化しつつ，これは簡単な例だよってことにしました．

```sh
$ mix deps.get
$ mix compile
$ iex -S mix
```

```elixir
iex()> {:ok, session} = Zenohex.open()
{:ok, #Reference<>}
iex()> {:ok, publisher} = Zenohex.Session.declare_publisher(session, "demo/example/test")
{:ok, #Reference<>}

> iex()> {:ok, subscriber} = Zenohex.Session.declare_subscriber(session, "demo/**")

Author: key名はいちばん使ってそうな Examples 例に合わせることにしました
https://github.com/eclipse-zenoh/zenoh-python/tree/master/examples#z_pub

{:ok, #Reference<>}
iex()> Zenohex.Publisher.put(publisher, "Hello Zenoh Dragon")
:ok
iex()> Zenohex.Subscriber.recv_timeout(subscriber, 1000)
{:ok,
 %Zenohex.Sample{
   key_expr: "demo/example/test",
   value: "Hello Zenoh Dragon",
   kind: :put,
   reference: #Reference<>

>  }}

Reviewer: あってます！途中で データそのまま（integer, float, binary）を返すだけでは設計が成り立たないことが分かって、Sample 構造体を返すようにしました。 私の範囲での README.md の修正し忘れです :bow:

iex()> Zenohex.Subscriber.recv_timeout(subscriber, 1000)
{:error, :timeout}
```

### Practical examples

We implemented practical examples under the [lib/zenohex/examples](https://github.com/b5g-ex/zenohex/tree/v0.2.0-rc.2/lib/zenohex/examples).
Since they consist of `Supervisor` and `GenServer`, we think they are useful as examples of more Elixir-like applications.

Please read the [lib/zenohex/examples/README.md](https://github.com/b5g-ex/zenohex/tree/v0.2.0-rc.2/lib/zenohex/examples/README.md) to use them as your implementation's reference.

## For developers

For most users, this section should be skipped.

> ### Build NIF module locally

Author: 自分でNIFコンパイルしたい方向けの情報はこのセクションにまとめました

This subsection is for developers who want to build NIF module locally in your Elixir application or try to use this repository itself for the contribution (very welcome!!).

First, please install and configure the Rust environment according to [the instructions on the official site](https://www.rust-lang.org/tools/install).

Then, add the following to your config file (e.g., `config/config.exs`) or make sure it is added.

```elixir
import Config

config :rustler_precompiled, :force_build, zenohex: true
```

When you want to build NIF module locally into your project, install Rustler by adding `rustler` to your list of dependencies in `mix.exs`:

```elixir
  defp deps do
    [
      ...
      {:zenohex, "~> 0.2.0-rc.2"},
      {:rustler, ">= 0.0.0", optional: true},
      ...
    ]
  end
```

> ### Enhance mix test

Author: `mix test`に関するTIPS的なところをここにまとめました．
いちばん自信がないので特に確認いただきたいです！ > @pojiro

Reviewer: ばっちり、問題なしです。

This subsection describes some Tips for `mix test`.

The default `mix test` rebuilds the NIF module, so you can reduce the test time by doing the following in advance.

```sh
export API_OPEN_SESSION_DELAY=0 && mix compile --force
```

You can also reduce the test time by adjusting `SCOUTING_DELAY` as the follow.

```sh
SCOUTING_DELAY=30 mix test
```

This parameter is used to set the upper time limit for searching (scouting) for a communication peer node.
IOW, if the test fails because the communication partner is not found within the set time, this value should be increased.

Finally, the default `mix test` only checks the communication of Zenoh nodes within the same session.
If you wish to run communication tests between different sessions, please run the following (CI does this in `test-with-another-session`).

```sh
USE_DIFFERENT_SESSION="1" mix test
```

### How to release

> These steps just follow the [Recommended flow of rustler_precompiled](https://hexdocs.pm/rustler_precompiled/precompilation_guide.html#recommended-flow).

Author: このセクションのバージョン番号は都度で更新したくないので，もう v0.2.0 ってことにしました．

1. Change versions, `mix.exs`, `native/zenohex_nif/Cargo.toml`
2. Run test, this step changes `native/zenohex_nif/Cargo.lock` version
3. Commit them and put the version tag, like v0.2.0

> 4. Puth the tag, like `git push origin v0.2.0`. this step triggers the `.github/workflows/nif_precompile.yml`

Reviewer: ```suggestion
4. Puth the tag, like `git push origin v0.2.0`. This step triggers the `.github/workflows/nif_precompile.yml`
```

スペルチェックでかかったので修正

> 5. After the artifacts are made, run `mix rustler_precompiled.download Zenohex.Nif --all` to update `checksum-Elixir.Zenohex.Nif.exs` and commit it.

Reviewer: ```suggestion
5. After the artifacts are made, run `mix rustler_precompiled.download Zenohex.Nif --all` to update `checksum-Elixir.Zenohex.Nif.exs` and commit it
```

他の項と合わせてピリオドなしにしました。（私が記述した際のゆらぎの修正

config :rustler_precompiled, :force_build, zenohex: true
```

When you want to build NIF module locally into your project, install Rustler by adding `rustler` to your list of dependencies in `mix.exs`:

```elixir
  defp deps do
    [
      ...
      {:zenohex, "~> 0.2.0-rc.2"},
      {:rustler, ">= 0.0.0", optional: true},
      ...
    ]
  end
```

### Enhance mix test

This subsection describes some Tips for `mix test`.

The default `mix test` rebuilds the NIF module, so you can reduce the test time by doing the following in advance.

```sh
export API_OPEN_SESSION_DELAY=0 && mix compile --force
```

You can also reduce the test time by adjusting `SCOUTING_DELAY` as the follow.

```sh
SCOUTING_DELAY=30 mix test
```

This parameter is used to set the upper time limit for searching (scouting) for a communication peer node.
IOW, if the test fails because the communication partner is not found within the set time, this value should be increased.

Finally, the default `mix test` only checks the communication of Zenoh nodes within the same session.
If you wish to run communication tests between different sessions, please run the following.

```sh
USE_DIFFERENT_SESSION="1" mix test
```

> FYI, CI does this in `test-with-another-session` without defining `SCOUTING_DELAY`, so this test sometimes fails on GHA.

Author: @pojiro CIが実際に落ちることがあるので [aaad576](https://github.com/b5g-ex/zenohex/pull/24/commits/aaad57687558f8a8ef9486088481757b103b8df3) にて注記しました（自分が忘れないために:D
https://github.com/b5g-ex/zenohex/actions/runs/7997846597/job/21843046901

Author: "without defining `SCOUTING_DELAY`" だと思っていましたが，合っていたでしょうか？

Reviewer: `USE_DIFFERENT_SESSION="1"` の場合、`SCOUTING_DELAY`  が短いより長い方が落ちにくくなるのは事実。
だけど、長くても（デフォルトの200ms としても）落ちる時があるというのが私の感触です。

Author: ありがとうございます！説明文としては間違っていない（`SCOUTING_DELAY`は定義していない）と認識しました．
ここで言いたかったことは `test-with-another-session` だけ落ちることあるけどトランキーロだぜっという自分たち向けのメモなので，不足ないということにします！

Author: なにかが足りてない,,, と思ったら `SCOUTING_DELAY` のデフォルト値が抜けてました．ので 78ac00b にて追加しました！

```

When you want to build NIF module locally into your project, install Rustler by adding `rustler` to your list of dependencies in `mix.exs`:

```elixir
  defp deps do
    [
      ...
      {:zenohex, "~> 0.2.0-rc.2"},
      {:rustler, ">= 0.0.0", optional: true},
      ...
    ]
  end
```

### Enhance mix test

This subsection describes some Tips for `mix test`.

The default `mix test` rebuilds the NIF module, so you can reduce the test time by doing the following in advance.

```sh
export API_OPEN_SESSION_DELAY=0 && mix compile --force
```

You can also reduce the test time by adjusting `SCOUTING_DELAY` as the follow.

```sh
SCOUTING_DELAY=30 mix test
```

This parameter is used to set the upper time limit for searching (scouting) for a communication peer node.
The default value (when undefined) is 200 ms.
IOW, if the test fails because the communication partner is not found within the set time, this value should be increased.

Finally, the default `mix test` only checks the communication of Zenoh nodes within the same session.
If you wish to run communication tests between different sessions, please run the following.

```sh
USE_DIFFERENT_SESSION="1" mix test
```

> FYI, CI does this in `test-with-another-session` without defining `SCOUTING_DELAY`, so this test sometimes fails on GHA.

Reviewer: ```suggestion
FYI, CI does this in `test-with-another-session` without defining `SCOUTING_DELAY`.
This test may fails on GHA, it depends on whether the scouting is successful or not.
```

落ちることもあるというニュアンスを残す文章にするならこうはどうでしょう？

Reviewer: 一足遅かった笑、🆗

Author: ごめんなさい！これ merge しちゃったのでこっちに移動させます！（もうひとつ修正漏れもあったので,,, 
https://github.com/b5g-ex/zenohex/pull/25
